src/consensus.rs

```diff
@@ -243,6 +221,88 @@ impl Consensus {
         }
     }
 
+    fn propose_updates(&mut self, timeout: Duration) -> anyhow::Result<()> {
+        match self.receiver.recv_timeout(timeout) {
+            Ok(Message::FromPeer(message)) => {
+                if message.get_msg_type() == MessageType::MsgHeartbeat
+                    || message.get_msg_type() == MessageType::MsgHeartbeatResponse
+                {
+                    log::trace!(
+                        "Received a message from peer with progress: {:?}. Message: {:?}",
+                        self.node.raft.prs().get(message.from),
+                        message
+                    );
+                } else {
+                    log::debug!(
+                        "Received a message from peer with progress: {:?}. Message: {:?}",
+                        self.node.raft.prs().get(message.from),
+                        message
+                    );
+                }
+                self.node.step(*message)?
+            }
+            Ok(Message::FromClient(message)) => {
+                log::debug!("Proposing entry from client with length: {}", message.len());
+                self.node.propose(vec![], message)?
+            }
+            Ok(Message::ConfChange(change)) => {
+                log::debug!("Proposing network configuration change: {:?}", change);
+                self.node.propose_conf_change(vec![], change)?
+            }
+            Err(RecvTimeoutError::Timeout) => (),
+            Err(RecvTimeoutError::Disconnected) => {
+                log::warn!("Stopping Raft as message sender was dropped");
+                return Ok(());
+            }
+        }
+        Ok(())
+    }
+
+    /// Returns `true` if learner promotion was proposed, `false` otherwise.
+    fn try_promote_learner(&mut self) -> anyhow::Result<bool> {
+        let store = self.node.store();
+        let commit = store.hard_state().commit;
+        let last_log_entry = store.last_index()?;
+        // Promote only when there are no uncommitted changes.
+        if commit != last_log_entry {
+            return Ok(false);
+        }
+        let status = self.node.status();
+        // Promote only if leader
+        if status.ss.raft_state != StateRole::Leader {
+            return Ok(false);
+        }
+        if let Some(learner) = self.find_learner_to_promote() {
```

Comment:
I think that in the steady state there will be zero learners most the time.
Maybe `self.find_learner_to_promote()` could be used first to shortcut the function right away if there are no learners to promote anyway.